### ISSUE TITLE
Add option to not read response into buffer, instead pass-through as stream

### ISSUE THREAD

**Issue by devinivy** [hapijs/shot#78]
I'm interested in receiving a response payload as a stream and handling that stream myself.  Currently the entire response payload is read into a buffer, which I'm trying to avoid in order to keep memory in check.  I'm happy to explain my use-case if necessary.

A nice property of this feature is that shot could then be used a perfect passthrough, if you imagine the case that a request comes to a live server, the server uses shot to resolve the response, then the server responds with a readable stream provided by shot.

Worth adding that I would hope to see this feature usable in hapi's `server.inject()`, so it's possible there's a higher-level conversation that needs to happen.

**Comment by hueniverse** [comment]
I am not ruling this out, but show is basically a hack around node's HTTP objects. It's playing games with the internal state to grab what it needs. I personally have no intention of revisiting this code and any change to it will require significant review because it has been a very stable hack for a long time. The most effective way to optimize this is to actually stop using shot and instead let hapi optimize the case where you are not actually using streams in the response. In those cases we don't need to go through all this serialization nonsense.

**Comment by devinivy** [comment]
Yeah, I see how this is a finicky area of the hapi codebase.  If we are interested at all in going down this path, ServerResponses do at least write to the socket as though it's just a duplex stream, and that "socket" may be set with `assignSocket()`– it appears it has worked that way for years.  The only reason we even get the `output` buffer currently used by shot is because no socket is assigned.  Food for thought.  It's possible this approach could lead to simplification within shot, even for shot's current features.  Curious what @mtharrison thinks.  I do agree it would require fairly significant review.

**Comment by mtharrison** [comment]
@devinivy what would that stream look like in your mind? Would it be (1) just the raw socket? So all the headers, CRLFs and protocol stuff? Would it just be (2) the response body as a stream, in which case you'd still want shot to split off headers and provide them in object with headers, statusCode etc.

Or would it be (3) a full instance of `http.IncomingMessage` as if you were the client?

Just trying to gauge expectations before going any deeper here.

**Comment by devinivy** [comment]
I would expect it to be the response body– basically a replacement for shot's `payload` and `rawPayload`.  I tried the `assignSocket()` approach out, and it does seem to work.

**Comment by hueniverse** [comment]
Still active?

**Comment by mtharrison** [comment]
@devinivy still something you think would be useful?

**Comment by devinivy** [comment]
I'm still into it and would be happy to help move it along or write some code– I realize it's a non-trivial amount of work.  I have my own purposes, but it would also be great if `server.inject()` supported it.  I guess if there's no interest there, it might not be as worthwhile.

**Comment by mtharrison** [comment]
@devinivy sure, I can't speak for whether it will be supported into `server.inject()` but I can see the value in it. Feel free to propose something. Sorry it's taken me a while to get back to you.

**Comment by devinivy** [comment]
I found some time to create a proof of concept.  I opted to keep things feeling similar, whether you're streaming or not.

`Shot.inject()` takes a new `stream` option that, when `true`, causes `Shot.inject()` to a resolve an object with identical properties to its typical return value (`raw`, `statusCode, `headers`, `trailers`, etc.).  But this object is also a readable for the response.  When the stream finishes it's properties are filled-in with all the final values (again, for `raw`, `statusCode, `headers`, `trailers`, etc.).

This isn't necessarily a proposed implementation– I think it would need some work– but I wanted to make some headway on this problem.

https://github.com/hapijs/shot/compare/master...devinivy:streaming
